### PR TITLE
Additive time signatures

### LINKED ISSUE #987: Support for additive time signatures
This is a feature request to support additive time signatures. For example, `.addTimeSignature('2+2+3/8')` would render a time signature like below:

Screen Shot 2021-05-17 at 22 34 09

### LINKED ISSUE #1129: Petaluma Font: Remove NOGLYPH glyphs
238 of the glyphs in `petaluma_glyphs.ts` have an outline that renders a [NOGLYPH] image.

Search for
```
m 1181 1440 l 268 1440 l 268 -1440 l 1181 -1440 z m 959 -524 l 959 -850 l 481 -850 l 481 -524 z m 864 -619 l 576 -619 l 576 -755 l 864 -755 z m 959 -26 l 959 -351 l 481 -351 l 481 -26 l 576 -26 l 576 -256 l 864 -256 l 864 -121 l 769 -121 l 769 -187 l 671 -187 l 671 -26 z m 959 325 l 959 37 l 481 37 l 481 135 l 864 135 l 864 325 z m 959 498 l 959 403 l 769 403 l 769 202 l 481 202 l 481 297 l 671 297 l 671 403 l 481 403 l 481 498 z m 671 778 l 576 778 l 576 674 l 671 674 z m 769 876 l 769 674 l 959 674 l 959 576 l 481 576 l 481 876 z m 959 1247 l 959 1152 l 769 1152 l 769 1045 l 959 1045 l 959 950 l 481 950 l 481 1045 l 671 1045 l 671 1152 l 484 1152 l 484 1247 z m 959 -916 l 959 -1011 l 867 -1011 l 665 -1146 l 959 -1146 l 959 -1241 l 481 -1241 l 481 -1146 l 683 -1011 l 481 -1011 l 481 -916 z
```

We should remove these glyphs, and then update the FONT_STACKS in vexflow_test_helpers to use Bravura as a fallback font for Petaluma. That way, when a glyph is missing in Petaluma, it just pulls the glyph from Bravura (which should include most of the glyphs we care about).

### PR DESCRIPTION
**Pull request by rvilarl**
Fixes #987
- Time Signatures now allow '+', '-', '(' and ')'

### CODE REVIEW

**Comment by rvilarl** [comment]
@mscuthbert the addtive and complex tests make probably no sense, can you look into that?
image

**Comment by rvilarl** [comment]
Issues:
- **Gonville** glyphs produced by `fontgen_gonville_custom.js` for '+', '(', ')' are wrong.
- A '+' defined by `.addTimeSignature('+')` is not centered in the stave.

**Comment by rvilarl** [comment]
After investigating the issue, I believe that it is a matter of setting the right vxx identification to the new glyphs in valid_codes. Can someone help on identifying the right codes for:
```
  timeSigPlus: 'vb8',
  timeSigPlusSmall: 'vbb',
  timeSigParensRight: 'vbc',
  timeSigParensRightSmall: 'vc0',
  timeSigParensLeft: 'vbc',
  timeSigParensRightSmall: 'vc1',
  brassScoop: 'vc4',
  brassDoitMedium: 'vc5',
  brassFallLipShort: 'vc6',
  brassLiftMedium: 'vc7',
  brassFallRoughMedium: 'vc8',
  brassBend: 'vc9',
  brassMuteClosed: 'vca',
  brassMuteOpen: 'vcb',
  brassFlip: 'vcc',
  brassJazzTurn: 'vcd',
  brassSmear: 'vce',
```
The current one vb8, ...are obviously wrong

**Comment by 0xfe** [comment]
Hi Rodriguo, any visual diffs from this one?

**Comment by rvilarl** [comment]
All the NOGLYPHS are gone! Then one or two instance with changes that cannot be appreciated.
I would say no visual differences, only the expected ones.

**Comment by rvilarl** [comment]
I have rebased to double check. No visual differences apart from the expected ones.

**Comment by rvilarl** [comment]
I have fixed grunt reference . No visual differences apart from the expected ones.

File: src/timesignature.ts [hunk 1]
```diff
@@ -18,18 +18,18 @@ export interface TimeSignatureInfo {
...
     throw new RuntimeError(
       'BadTimeSignature',
       `Invalid time spec: ${timeSpec}. Must be in the form "<numerator>/<denominator>"`
     );
   }
 
   numbers.forEach((number) => {
-    if (isNaN(Number(number))) {
-      throw new RuntimeError('BadTimeSignature', `Invalid time spec: ${timeSpec}. Must contain two valid numbers.`);
+    if (/^([0-9+()]{1,})$/.test(number) == false) {
```
Comment (0xfe) [inline src/timesignature.ts:31]: Can you add a comment here that makes the format explicit. (Better still, I'd suggest not using a regex for this simple case.)

File: src/timesignature.ts [hunk 2]
```diff
@@ -18,18 +18,18 @@ export interface TimeSignatureInfo {
 }
 
 const assertIsValidFraction = (timeSpec: string) => {
```
Comment (0xfe) [inline src/timesignature.ts:20]: Maybe `assertIsValidTimeSig` since the implementation is different now

File: src/timesignature.ts [hunk 3]
```diff
@@ -92,15 +92,15 @@ export class TimeSignature extends StaveModifier {
       assertIsValidFraction(timeSpec);
     }
 
-    const [topDigits, botDigits] = timeSpec.split('/').map((number) => number.split(''));
+    const parts = timeSpec.split('/');
 
     return {
       num: true,
-      glyph: this.makeTimeSignatureGlyph(topDigits, botDigits),
+      glyph: this.makeTimeSignatureGlyph(parts[0] ?? '', parts[1] ?? ''),
```
Comment (0xfe) [inline src/timesignature.ts:99]: The `?? ''` look like silent failures here. Shouldn't the assertion above throw if the parts are going to be empty?

File: tests/timesignature_tests.ts [hunk 4]
```diff
@@ -17,6 +17,8 @@ const TimeSignatureTests = {
     const run = VexFlowTests.runTests;
     run('Basic Time Signatures', basic);
     run('Big Signature Test', big);
+    run('Additive Signature Test', additive);
```
Comment (0xfe) [inline tests/timesignature_tests.ts:20]: Can we add a "Bad Signatures Test" to verify the corner cases of the parsing?

File: tests/timesignature_tests.ts [hunk 5]
```diff
@@ -73,6 +75,27 @@ function big(options: TestOptions, contextBuilder: ContextBuilder): void {
...
+  ok(true, 'all pass');
+}
+
+function complex(options: TestOptions, contextBuilder: ContextBuilder): void {
+  const ctx = contextBuilder(options.elementId, 400, 120);
+
+  new Stave(10, 10, 300)
+    .addTimeSignature('(2+3)/16')
+    .addTimeSignature('+')
+    .addTimeSignature('3/8')
```
Comment (0xfe) [inline tests/timesignature_tests.ts:92]: Can we add a few more different combinations here?

**Review by 0xfe — COMMENTED** [review]
Looks good overall, Rodrigo. Some minor comments around testing corner cases because the format is a bit more complex now. (Also if you could separate the 'No glyphs' that would be great.

File: src/timesignature.ts (hunk 1 above)
Comment (rvilarl) [reply src/timesignature.ts:31]: How would you make the test then? Comment would be:
"One or more characters consisting in number 0..9, '+', '(' or ')'"

File: src/timesignature.ts (hunk 3 above)
Comment (rvilarl) [reply src/timesignature.ts:99]: Now it is possible to have only one part in the case of '+'

File: tests/timesignature_tests.ts (hunk 4 above)
Comment (rvilarl) [reply tests/timesignature_tests.ts:20]: Could you suggest some cases that you would like to see?

File: tests/timesignature_tests.ts (hunk 5 above)
Comment (rvilarl) [reply tests/timesignature_tests.ts:92]: Of course but I would like them to make sense. This is why I asked @mscuthbert. I will try to find some examples in internet.

File: src/timesignature.ts (hunk 1 above)
Comment (0xfe) [reply src/timesignature.ts:31]: That comment is fine. Maybe simpler regex: `/^[0-9+()]+$/` for the filter.

But for a valid signature, I'd imagine, you need two tests:

* without parens: `/^[0-9]+(\+[0-9]+)?`
* open paren must close: `/^\([0-9]+(\+[0-9]+)?\)\)$/`

File: src/timesignature.ts (hunk 3 above)
Comment (0xfe) [reply src/timesignature.ts:99]: See my comment about the regex test above. (If I understand your comment correctly, you're saying this accepts `4+` when it should reject it, right?)

File: tests/timesignature_tests.ts (hunk 4 above)
Comment (0xfe) [reply tests/timesignature_tests.ts:20]: `+`, `1+`, `+1`, `(3+` , `+3)`, `()`, `(+)`, `(4/5)`, `(4+3/)5`, etc.

Lots of edge cases here.

File: tests/timesignature_tests.ts (hunk 5 above)
Comment (0xfe) [reply tests/timesignature_tests.ts:92]: Gotcha. Is `+` supposed to be valid?

Comment (rvilarl) [reply tests/timesignature_tests.ts:92]: yes, it is supposed to be valid. If we do not structure the complex signatures like that, parsing them would be really complex. The example above is one signature that corresponds to the complex one that @mscuthbert provided in #987.

File: src/timesignature.ts (hunk 1 above)
Comment (rvilarl) [reply src/timesignature.ts:31]: I will go for the simple one, if someone does not close the parenthesis, they will see that in the results.

File: src/timesignature.ts (hunk 2 above)
Comment (rvilarl) [reply src/timesignature.ts:20]: Done

File: src/timesignature.ts (hunk 3 above)
Comment (rvilarl) [reply src/timesignature.ts:99]: No, I am syaing that this accepts `+` this is done as described below to avoid parsing complex signatures.

File: tests/timesignature_tests.ts (hunk 4 above)
Comment (rvilarl) [reply tests/timesignature_tests.ts:20]: I have included this cases in the parser test.
- `x` is valid
- `(4/5)` `(4+3/)5` are also valid using the simple regex. I think that this is no problem as the user will identify that the rendered result is not right. I left them out.

**Comment by rvilarl** [comment]
@0xfe please merge first #1248 to reduce the number of changes in this PR. 
Comments addressed ready to merge.

**Comment by rvilarl** [comment]
@0xfe I have decided to include the examples mentioned in this page:
https://steinberg.help/dorico/v3/en/dorico/topics/notation_reference/notation_reference_time_signatures/notation_reference_time_signatures_types_c.html

With this I have seen that the `-` was also necessary.

**Comment by rvilarl** [comment]
**TimeSignature Additive_Signature_Test Bravura**
TimeSignature Additive_Signature_Test Bravura
**TimeSignature Additive_Signature_Test Gonville**
TimeSignature Additive_Signature_Test Gonville
**TimeSignature Additive_Signature_Test Petaluma**
TimeSignature Additive_Signature_Test Petaluma

**Comment by rvilarl** [comment]
**TimeSignature Aggregate_Signature_Test Bravura**
TimeSignature Aggregate_Signature_Test Bravura
**TimeSignature Aggregate_Signature_Test Gonville**
TimeSignature Aggregate_Signature_Test Gonville
**TimeSignature Aggregate_Signature_Test Petaluma**
TimeSignature Aggregate_Signature_Test Petaluma

**Comment by rvilarl** [comment]
**TimeSignature Alternating_Signature_Test Bravura**
TimeSignature Alternating_Signature_Test Bravura
**TimeSignature Alternating_Signature_Test Gonville**
TimeSignature Alternating_Signature_Test Gonville
**TimeSignature Alternating_Signature_Test Petaluma**
TimeSignature Alternating_Signature_Test Petaluma

**Comment by rvilarl** [comment]
**TimeSignature Interchangeable_Signature_Test Bravura**
TimeSignature Interchangeable_Signature_Test Bravura
**TimeSignature Interchangeable_Signature_Test Gonville**
TimeSignature Interchangeable_Signature_Test Gonville
**TimeSignature Interchangeable_Signature_Test Petaluma**
TimeSignature Interchangeable_Signature_Test Petaluma

**Comment by rvilarl** [comment]
and the complex example from mscuthbert 
**TimeSignature Complex_Signature_Test Bravura**
TimeSignature Complex_Signature_Test Bravura
**TimeSignature Complex_Signature_Test Gonville**
TimeSignature Complex_Signature_Test Gonville
**TimeSignature Complex_Signature_Test Petaluma**
TimeSignature Complex_Signature_Test Petaluma

**Comment by rvilarl** [comment]
I found a more adequate glyph for the `+` and `-`, however the parenthesis are still moved down. 
I would say that there is too much space around `+` and `-` when placed in the middle making an additional call to `addTimeSignature`
This two topics could be solved later on by @AaronDavidNewman, he is an expert on relocation of glyphs.

I would say ready to review and merge.

File: tests/timesignature_tests.ts (hunk 5 above)
Comment (rvilarl) [reply tests/timesignature_tests.ts:126]: Examples added.

**Comment by rvilarl** [comment]
I have commited  the gonville inspector that I used to identify the glyphs.
Ready to review.

**Comment by rvilarl** [comment]
@0xfe I think that there is no open topic here. Please merge #1232 first to avoid conflicts.

**Comment by 0xfe** [comment]
Thanks for the extra tests, Rodrigo. I think this looks good.
